### PR TITLE
Add protobuf-c/1.4.0

### PR DESCRIPTION
**Pull request by bemehiser**
protobuf-c/1.4.0

Add Google Protobuf-C Library

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/protobuf-c/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    def package(self):
+        self.copy("LICENSE", src=self._source_subfolder, dst="licenses", keep_path=False)
+
+    def package_info(self):
+        self.cpp_info.filenames["cmake_find_package"] = "ProtobufC"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "protobuf-c"
+        self.cpp_info.names["cmake"] = "protobuf-c"
+        self.cpp_info.names["cmake_paths"] = "protobuf-c"
+        self.cpp_info.names["cmake_find_package"] = "protobuf-c"
+        self.cpp_info.names["cmake_find_package_multi"] = "protobuf-c"
```
Comment (SpaceIm) [inline recipes/protobuf-c/all/conanfile.py:65]: Where does it come from? Upstream doesn't export target in a config file.
`self.cpp_info.names["cmake"]` and `self.cpp_info.names["cmake_paths"]` don't make sense for conan AFAIK.

File: recipes/protobuf-c/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        cmake.definitions["BUILD_PROTO3"] = self.options.build_proto3
+        cmake.definitions["BUILD_PROTOC"] = self.options.build_protoc
+        cmake.definitions["BUILD_SHARED_LIBS"] = self.options.shared
+        cmake.definitions["BUILD_TESTS"] = False
+        cmake.definitions["HAS_SYSTEM_PROTOBUF"] = False
+        cmake.definitions["CMAKE_VERBOSE_MAKEFILE"] = self.options.verbose_makefile
+        cmake.verbose = True
+        cmake.configure(source_folder=f"{self._source_subfolder}/build-cmake")
+        cmake.build()
+        cmake.install()
```
Comment (SpaceIm) [inline recipes/protobuf-c/all/conanfile.py:54]: installation must occur in `package()`

File: recipes/protobuf-c/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            if self.settings.compiler.cppstd:
+                tools.check_min_cppstd(self, 11)
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename(f"{self.name}-{self.version}", self._source_subfolder)
+
+        if "patches" in self.conan_data and self.version in self.conan_data["patches"]:
+            for patch in self.conan_data["patches"][self.version]:
+                tools.patch(**patch)
```
Comment (SpaceIm) [inline recipes/protobuf-c/all/conanfile.py:41]: do not patch in `source()`

File: recipes/protobuf-c/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        "build_proto3": True,
+        "build_protoc": True,
+        "shared": False,
+        "verbose_makefile": False,
+    }
+
+    def config_options(self):
+        if self.settings.compiler == 'gcc' and float(self.settings.compiler.version.value) >= 5.1:
+            if self.settings.compiler.cppstd:
+                tools.check_min_cppstd(self, 11)
```
Comment (SpaceIm) [inline recipes/protobuf-c/all/conanfile.py:33]: - this logic must go into `validate()`
- this is not how you can compare versions.
- why gcc >= 5.1 needs C++11 but not other compilers?

File: recipes/protobuf-c/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    exports_sources = ["patches/**"]
+    _source_subfolder = "source_subfolder"
+    requires = [
+        "protobuf/3.19.2",
+    ]
+    options = {
+        "build_proto3": [True, False],
+        "build_protoc": [True, False],
+        "shared": [True, False],
+        "verbose_makefile": [True, False],
```
Comment (SpaceIm) [inline recipes/protobuf-c/all/conanfile.py:21]: - add fPIC option please
- try to follow the same order than all other CCI recipes: shared, fPIC, then other options
- why `verbose_makefile` option? We don't care unless I miss something. It's just noise in recipe options.
- I think you can remove `build_` in other options, so proto3 & protoc.

**Comment by bemehiser** [comment]
@SpaceIm, thank you for the detailed review!

**Comment by SpaceIm** [comment]
> @SpaceIm, thank you for the detailed review!
> 
> You don't happen to know how to make conan-center-index only build the c++11 configurations, do you? protobuf-c needs c++11, and doesn't work with older versions.

It should be handled by upstream CMake files. If not, you have to patch them.

**Review by gegles — APPROVED** [review]
Thanks @bemehiser!

File: recipes/protobuf-c/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
+import functools
+
+from conans import ConanFile, CMake, tools
+import os
+
+
+class ProtobufcConan(ConanFile):
+    name = "protobuf-c"
+    license = "https://github.com/protobuf-c/protobuf-c/blob/master/LICENSE"
```
Comment (prince-chrismc) [inline recipes/protobuf-c/all/conanfile.py:9]: https://github.com/conan-io/conan-center-index/blob/master/docs/reviewing.md#license-attribute

Looks like a modified https://spdx.org/licenses/BSD-2-Clause

```suggestion
    license = "LicenseRef-LICENSE"
```

File: recipes/protobuf-c/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+
+    def validate(self):
+        tools.check_min_cppstd(self, "11")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        os.rename(f"{self.name}-{self.version}", self._source_subfolder)
```
Comment (prince-chrismc) [inline recipes/protobuf-c/all/conanfile.py:47]:
```suggestion
        tools.get(self, **self.conan_data["sources"][self.version], destination=self._source_subfolder, strip_root=True)
```

File: recipes/protobuf-c/all/patches/1.4.0-cmake-remove-pkgconfig.patch [hunk 8]
```diff
@@ -0,0 +1,13 @@
+diff --git a/build-cmake/CMakeLists.txt b/build-cmake/CMakeLists.txt
+index d4d015c..f0e6df0 100644
+--- a/build-cmake/CMakeLists.txt
++++ b/build-cmake/CMakeLists.txt
+@@ -199,8 +199,6 @@ SET(exec_prefix \${prefix})
+ SET(bindir \${exec_prefix}/${CMAKE_INSTALL_BINDIR})
+ SET(libdir \${exec_prefix}/${CMAKE_INSTALL_LIBDIR})
+ SET(includedir \${prefix}/${CMAKE_INSTALL_INCLUDEDIR})
+-CONFIGURE_FILE(${MAIN_DIR}/protobuf-c/libprotobuf-c.pc.in libprotobuf-c.pc @ONLY)
+-INSTALL(FILES ${CMAKE_BINARY_DIR}/libprotobuf-c.pc DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
```
Comment (prince-chrismc) [inline recipes/protobuf-c/all/patches/1.4.0-cmake-remove-pkgconfig.patch:10]: You should also set the pkg_config names [docs here](https://github.com/conan-io/conan-center-index/blob/master/docs/v2_migration.md#pkgconfigdeps)

File: recipes/protobuf-c/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    def package(self):
+        cmake = self._configure_cmake()
+        cmake.install()
+        self.copy("LICENSE", src=self._source_subfolder, dst="licenses", keep_path=False)
+
+    def package_info(self):
+        self.cpp_info.filenames["cmake_find_package"] = "ProtobufC"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "protobuf-c"
+        self.cpp_info.names["cmake_find_package"] = "protobuf-c"
+        self.cpp_info.names["cmake_find_package_multi"] = "protobuf-c"
```
Comment (prince-chrismc) [inline recipes/protobuf-c/all/conanfile.py:76]: https://github.com/conan-io/conan-center-index/blob/master/docs/v2_migration.md#translating-names-information-to-cmake_target_name-cmake_module_target_name-and-cmake_file_name

```suggestion
        self.cpp_info.names["cmake_find_package"] = "protobuf-c"
        self.cpp_info.names["cmake_find_package_multi"] = "protobuf-c"
        self.cpp_info.set_property("cmake_target_name", protobuf-c::protobuf-c")
```

File: recipes/protobuf-c/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        cmake.build()
+
+    def package(self):
+        cmake = self._configure_cmake()
+        cmake.install()
+        self.copy("LICENSE", src=self._source_subfolder, dst="licenses", keep_path=False)
+
+    def package_info(self):
+        self.cpp_info.filenames["cmake_find_package"] = "ProtobufC"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "protobuf-c"
```
Comment (prince-chrismc) [inline recipes/protobuf-c/all/conanfile.py:74]: FYI the `_multi` name looks out of place but I did not check if it's correct

***

More v2 suggestion

```suggestion
        self.cpp_info.filenames["cmake_find_package"] = "ProtobufC"
        self.cpp_info.filenames["cmake_find_package_multi"] = "protobuf-c"
        self.cpp_info.set_property("cmake_file_name", "ProtobufC")
```

**Pull request closed without merging**